mitodl/micromasters
# Added ability to edit employment on profile page

#### What are the relevant tickets?

closes #393

#### What's this PR do?

This PR makes some changes to enable editing profile fields on the `/users/:username` page. Doing this entailed the following changes:

- A new class, `EmploymentForm`, which actually draws the form portion of the employment history editing UI
- A new class, `EmploymentValidation`, which simply defines the values used for validation (some `propTypes`, `defaultProps`, and the `validation` class method).
- changes to the `ProfileTab`. Now instead of a class that all the tabs inherit from, it's a mixin (called `ProfileFormFields`) which provides bindings (on `this`) for all of our input field types. I didn't really want to have `EmploymentTab extends EmploymentValidation` and `EmploymentValidation extends ProfileTab` (and `ProfileTab extends React.Component`) which is the reason for this change.
- A new class, `ProfileFormContainer`, which is a utility class that defines helpers for dispatching actions and for injecting props into `this.props.children`. Basically, the idea here is that `ProfilePage` and `UserPage` inherit from `ProfileFormContainer`, and then are connected to the store. `ProfileFormContainer` basically makes it easier to do the right thing in two places.

the diff looks a lot bigger than it really is, this is mostly moving around existing code to different places.

#### Where should the reviewer start?

I would start by reviewing the changes to `EmploymentTab` and `EmploymentForm`.

#### How should this be manually tested?

Make sure there are no regressions on the profile. You should be able to edit fields and so on everywhere.

Also, visit `/users/:username` and make sure you can edit entries in the work history form. things should validate and so on as normal.

#### Any background context you want to provide?

...whew

#### Screenshots (if appropriate)

I didn't do any ui / pixel polish work in this PR, I think in one of the subsequent ones I'll try to make things look pretty, but here it is:

![users_things](https://cloud.githubusercontent.com/assets/6207644/15519146/077440bc-21ce-11e6-8f0b-acb722886ccc.png)


#### What GIF best describes this PR or how it makes you feel?
